File: src/test/java/com/heekng/celloct/repository/JoinRequestRepositoryTest.java
package com.heekng.celloct.repository;

import com.heekng.celloct.entity.JoinRequest;
import com.heekng.celloct.entity.Member;
import com.heekng.celloct.entity.Shop;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
@Transactional
class JoinRequestRepositoryTest {

    @Autowired
    EntityManager em;
    @Autowired
    ShopRepository shopRepository;
    @Autowired
    MemberRepository memberRepository;
    @Autowired
    JoinRequestRepository joinRequestRepository;

    @Test
    void create() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();
        //when
        JoinRequest findJoinRequest = joinRequestRepository.findById(joinRequest.getId()).get();
        //then
        assertThat(joinRequest.getMember().getId()).isEqualTo(findJoinRequest.getMember().getId());
        assertThat(joinRequest.getShop().getId()).isEqualTo(findJoinRequest.getShop().getId());
    }

    @Test
    void findByMemberId() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();

        //when
        List<JoinRequest> findJoinRequests = joinRequestRepository.findByMemberId(member1.getId());
        //then
        findJoinRequests.forEach(findJoinRequest -> assertThat(findJoinRequest.getMember().getId()).isEqualTo(member1.getId()));
    }

    @Test
    void findByShopId() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();
        //when
        List<JoinRequest> findJoinRequests = joinRequestRepository.findByShopId(shop.getId());
        //then
        findJoinRequests.forEach(findJoinRequest -> assertThat(findJoinRequest.getShop().getId()).isEqualTo(shop.getId()));
    }

    @Test
    void deleteJoinRequest() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();
        //when
        JoinRequest findJoinRequest = joinRequestRepository.findById(joinRequest.getId()).get();
        joinRequestRepository.delete(findJoinRequest);
        Optional<JoinRequest> joinRequestOptional = joinRequestRepository.findById(joinRequest.getId());
        //then
        assertThat(joinRequestOptional).isEmpty();
    }

    @Test
    void findByMemberIdAndShopIdTest() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();
        //when
        List<JoinRequest> joinRequests = joinRequestRepository.findByMemberIdAndShopId(member1.getId(), shop.getId());

        //then
        assertThat(joinRequests.get(0).getShop().getId()).isEqualTo(shop.getId());
        assertThat(joinRequests.get(0).getMember().getId()).isEqualTo(member1.getId());
    }

    @Test
    void findWithMemberByShopIdTest() throws Exception {
        //given
        Member member1 = Member.builder()
                .name("member1")
                .email("member1@test.com")
                .build();
        memberRepository.save(member1);

        Shop shop = Shop.builder()
                .phone("010-1234-1234")
                .name("shop1")
                .build();
        shopRepository.save(shop);

        JoinRequest joinRequest = JoinRequest.builder()
                .member(member1)
                .shop(shop)
                .build();
        joinRequestRepository.save(joinRequest);

        em.flush();
        em.clear();
        //when
        List<JoinRequest> joinRequests = joinRequestRepository.findWithMemberByShopId(shop.getId());
        em.flush();
        em.clear();

        //then
        assertThat(joinRequests.get(0).getMember().getId()).isEqualTo(member1.getId());
        assertThat(joinRequests.get(0).getShop().getId()).isEqualTo(shop.getId());
    }

}

File: src/test/kotlin/com/heekng/celloct/repository/JoinRequestRepositoryTest.kt
package com.heekng.celloct.repository

import com.heekng.celloct.entity.JoinRequest
import com.heekng.celloct.entity.Member
import com.heekng.celloct.entity.Shop
import com.heekng.celloct.util.findByIdOrThrow
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.data.repository.findByIdOrNull
import org.springframework.transaction.annotation.Transactional
import javax.persistence.EntityManager

@SpringBootTest
@Transactional
class JoinRequestRepositoryTest @Autowired constructor(
    val em: EntityManager,
    val shopRepository: ShopRepository,
    val memberRepository: MemberRepository,
    val joinRequestRepository: JoinRequestRepository,
) {

    @Test
    fun createTest() {
        //given
        val member = Member.fixture("member1", "member1@test.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
        val findJoinRequest = joinRequestRepository.findByIdOrThrow(joinRequest.id)
        //then
        assertThat(findJoinRequest.id).isEqualTo(joinRequest.id)
        assertThat(findJoinRequest.member.id).isEqualTo(member.id)
        assertThat(findJoinRequest.shop.id).isEqualTo(shop.id)
    }

    @Test
    fun findByMemberIdTest() {
        //given
        val member = Member.fixture("member", "member1@Test.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
        val findJoinRequests = joinRequestRepository.find(memberId = member.id)
        //then
        assertThat(findJoinRequests).extracting("member").extracting("id").containsExactlyInAnyOrder(member.id)
    }

    @Test
    fun findByShopIdTest() {
        //given
        val member1 = Member.fixture("member1", "member1@Test.com")
        memberRepository.save(member1)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member1, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
//        val findJoinRequests = joinRequestRepository.findByShopId(shop.id!!)
        val findJoinRequests = joinRequestRepository.find(shopId = shop.id)
        //then
        assertThat(findJoinRequests).extracting("shop").extracting("id").containsExactlyInAnyOrder(shop.id)
    }

    @Test
    fun deleteJoinRequestTest() {
        //given
        val member = Member.fixture("member1", "member1@Test.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
        val findJoinRequest = joinRequestRepository.findByIdOrThrow(joinRequest.id)
        joinRequestRepository.delete(findJoinRequest)
        val joinRequestOrNull = joinRequestRepository.findByIdOrNull(joinRequest.id)
        //then
        assertThat(joinRequestOrNull).isNull()
    }

    @Test
    fun findByMemberIdAndShopIdTest() {
        //given
        val member = Member.fixture("member1", "member1@test.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
        val findJoinRequests = joinRequestRepository.find(
            memberId = member.id,
            shopId = shop.id,
        )
        //then
        assertThat(findJoinRequests).hasSize(1)
        assertThat(findJoinRequests[0].member.id).isEqualTo(member.id)
        assertThat(findJoinRequests[0].shop.id).isEqualTo(shop.id)
    }

    @Test
    fun findWithMemberByShopIdTest() {
        //given
        val member = Member.fixture("member1", "member1@test.com")
        memberRepository.save(member)
        val shop = Shop.fixture("010-1234-1234", "shop1", null)
        shopRepository.save(shop)
        val joinRequest = JoinRequest.fixture(member, shop)
        joinRequestRepository.save(joinRequest)
        em.flush()
        em.clear()
        //when
        val findJoinRequests = joinRequestRepository.findWithMemberByShopId(shop.id!!)
        em.flush()
        em.clear()
        //then
        assertThat(findJoinRequests).hasSize(1)
        assertThat(findJoinRequests[0].member.id).isEqualTo(member.id)
        assertThat(findJoinRequests[0].shop.id).isEqualTo(shop.id)
    }

}